File: ui_components/intro_manager.py
"""
Intro/Tour Manager for ZEN_RAG
Shows an interactive first-run experience explaining app capabilities.
"""

import streamlit as st


class IntroTour:
    """Manages the intro/onboarding tour for new users."""

    # Session state key
    STATE_KEY = "intro_completed"
    SLIDE_KEY = "intro_slide"

    # Tour slides
    SLIDES = [
        {
            "title": "🔬 Welcome to ZEN_RAG",
            "subtitle": "Chat with your Documents",
            "content": """
            **ZEN_RAG** is a powerful Retrieval-Augmented Generation (RAG) system
            that lets you:
            
            • 📄 Upload and analyze documents (PDFs, images, text)
            • 🔍 Search across your document collection
            • 💬 Chat with your documents using AI
            • 🔐 Keep everything private with local models
            
            Let's explore the key features!
            """,
            "icon": "🔬",
        },
        {
            "title": "⚡ Hardware Aware",
            "subtitle": "Optimized for Your System",
            "content": """
            ZEN_RAG automatically detects and optimizes for your hardware:
            
            • **CPU**: Uses all available cores for parallel processing
            • **GPU**: Detects NVIDIA/AMD GPUs and offloads inference
            • **RAM**: Adaptive batch sizing based on available memory
            • **NPU**: Leverages Ryzen AI neural processors when available
            
            Your system is automatically configured for maximum performance!
            """,
            "icon": "⚡",
        },
        {
            "title": "🔒 Privacy Focused",
            "subtitle": "Your Data Stays Local",
            "content": """
            All processing happens on your machine:
            
            • **Local Models**: Run open-source LLMs via llama.cpp or Ollama
            • **No Cloud Dependency**: No data sent to external services
            • **Optional Cloud**: Use OpenAI, Claude, or Gemini if you prefer
            • **Full Control**: You decide what data to process and how
            
            Your privacy is guaranteed!
            """,
            "icon": "🔒",
        },
        {
            "title": "🚀 Getting Started",
            "subtitle": "Ready to Begin?",
            "content": """
            ### Next Steps:
            
            1. **Sidebar Settings**: Configure your LLM provider and model
            2. **Upload Documents**: Add PDFs or images to analyze
            3. **Chat**: Ask questions about your documents
            4. **Explore**: Check out advanced features in the sidebar
            
            ### Need Help?
            - See "Local Server Helper" to start llama-server
            - Visit the GitHub repo for documentation
            - Check out example documents in the sidebar
            
            Let's get started! ✨
            """,
            "icon": "🚀",
        },
    ]

    @staticmethod
    def should_show_intro() -> bool:
        """Check if intro should be shown."""
        return not st.session_state.get(IntroTour.STATE_KEY, False)

    @staticmethod
    def mark_intro_completed() -> None:
        """Mark intro as completed."""
        st.session_state[IntroTour.STATE_KEY] = True

    @staticmethod
    def reset_intro() -> None:
        """Reset intro (for testing or user request)."""
        st.session_state[IntroTour.STATE_KEY] = False
        st.session_state[IntroTour.SLIDE_KEY] = 0

    @staticmethod
    def show_intro_modal() -> None:
        """Display the intro tour as a modal overlay."""
        if not IntroTour.should_show_intro():
            return

        # Initialize slide counter
        if IntroTour.SLIDE_KEY not in st.session_state:
            st.session_state[IntroTour.SLIDE_KEY] = 0

        current_slide = st.session_state[IntroTour.SLIDE_KEY]
        slide = IntroTour.SLIDES[current_slide]

        # Create modal container
        with st.container():
            col1, col2, col3 = st.columns([1, 10, 1])

            with col2:
                st.markdown("---")

                # Header with icon
                st.markdown(
                    f"<h1 style='text-align: center'>{slide['icon']}</h1>",
                    unsafe_allow_html=True,
                )
                st.markdown(
                    f"<h2 style='text-align: center'>{slide['title']}</h2>",
                    unsafe_allow_html=True,
                )
                st.markdown(
                    f"<h3 style='text-align: center; color: gray'>{slide['subtitle']}</h3>",
                    unsafe_allow_html=True,
                )

                st.markdown("---")

                # Content
                st.markdown(slide["content"])

                st.markdown("---")

                # Navigation buttons
                col_prev, col_counter, col_next, col_skip = st.columns(4)

                with col_prev:
                    if current_slide > 0:
                        if st.button("← Previous", use_container_width=True):
                            st.session_state[IntroTour.SLIDE_KEY] -= 1
                            st.rerun()
                    else:
                        st.button("← Previous", use_container_width=True, disabled=True)

                with col_counter:
                    st.markdown(
                        f"<div style='text-align: center; padding-top: 10px'>"
                        f"<strong>{current_slide + 1} / {len(IntroTour.SLIDES)}</strong>"
                        f"</div>",
                        unsafe_allow_html=True,
                    )

                with col_next:
                    if current_slide < len(IntroTour.SLIDES) - 1:
                        if st.button("Next →", use_container_width=True):
                            st.session_state[IntroTour.SLIDE_KEY] += 1
                            st.rerun()
                    else:
                        if st.button(
                            "Finish 🎉",
                            use_container_width=True,
                            help="Complete the tour and start using ZEN_RAG",
                        ):
                            IntroTour.mark_intro_completed()
                            st.rerun()

                with col_skip:
                    if st.button("Skip", use_container_width=True):
                        IntroTour.mark_intro_completed()
                        st.rerun()

                st.markdown("---")


def show_intro_in_sidebar():
    """Show intro management controls in sidebar (for testing/reset)."""
    with st.sidebar:
        with st.expander("ℹ️ Help & Info", expanded=False):
            st.caption("**Intro Tour**")

            if st.session_state.get(IntroTour.STATE_KEY, False):
                if st.button("🔄 Show Intro Again", use_container_width=True):
                    IntroTour.reset_intro()
                    st.rerun()
            else:
                st.info("Intro tour will display on next visit.")

            st.markdown("---")
            st.caption("📚 **Resources**")
            st.markdown("""
            - [GitHub Repository](https://github.com/GeoHaber/ZEN_RAG)
            - [Documentation](https://github.com/GeoHaber/ZEN_RAG#readme)
            - [Issues & Feedback](https://github.com/GeoHaber/ZEN_RAG/issues)
            """)

File: rust_output/src/intro_manager.rs
/// Intro/Tour Manager for ZEN_RAG
/// Shows an interactive first-run experience explaining app capabilities.

use anyhow::{Result, Context};

/// Manages the intro/onboarding tour for new users.
#[derive(Debug, Clone)]
pub struct IntroTour {
}

impl IntroTour {
    /// Check if intro should be shown.
    pub fn should_show_intro() -> bool {
        // Check if intro should be shown.
        !streamlit.session_state.get(&IntroTour.STATE_KEY).cloned().unwrap_or(false)
    }
    /// Mark intro as completed.
    pub fn mark_intro_completed() -> () {
        // Mark intro as completed.
        streamlit.session_state[IntroTour.STATE_KEY] = true;
    }
    /// Reset intro (for testing or user request).
    pub fn reset_intro() -> () {
        // Reset intro (for testing or user request).
        streamlit.session_state[IntroTour.STATE_KEY] = false;
        streamlit.session_state[IntroTour.SLIDE_KEY] = 0;
    }
    /// Display the intro tour as a modal overlay.
    pub fn show_intro_modal() -> () {
        // Display the intro tour as a modal overlay.
        if !IntroTour.should_show_intro() {
            return;
        }
        if !streamlit.session_state.contains(&IntroTour.SLIDE_KEY) {
            streamlit.session_state[IntroTour.SLIDE_KEY] = 0;
        }
        let mut current_slide = streamlit.session_state[&IntroTour.SLIDE_KEY];
        let mut slide = IntroTour.SLIDES[&current_slide];
        let _ctx = streamlit.container();
        {
            let (mut col1, mut col2, mut col3) = streamlit.columns(vec![1, 10, 1]);
            let _ctx = col2;
            {
                streamlit.markdown("---".to_string());
                streamlit.markdown(format!("<h1 style='text-align: center'>{}</h1>", slide["icon".to_string()]), /* unsafe_allow_html= */ true);
                streamlit.markdown(format!("<h2 style='text-align: center'>{}</h2>", slide["title".to_string()]), /* unsafe_allow_html= */ true);
                streamlit.markdown(format!("<h3 style='text-align: center; color: gray'>{}</h3>", slide["subtitle".to_string()]), /* unsafe_allow_html= */ true);
                streamlit.markdown("---".to_string());
                streamlit.markdown(slide["content".to_string()]);
                streamlit.markdown("---".to_string());
                let (mut col_prev, mut col_counter, mut col_next, mut col_skip) = streamlit.columns(4);
                let _ctx = col_prev;
                {
                    if current_slide > 0 {
                        if streamlit.button("← Previous".to_string(), /* use_container_width= */ true) {
                            streamlit.session_state[IntroTour.SLIDE_KEY] -= 1;
                            streamlit.rerun();
                        }
                    } else {
                        streamlit.button("← Previous".to_string(), /* use_container_width= */ true, /* disabled= */ true);
                    }
                }
                let _ctx = col_counter;
                {
                    streamlit.markdown(format!("<div style='text-align: center; padding-top: 10px'><strong>{} / {}</strong></div>", (current_slide + 1), IntroTour.SLIDES.len()), /* unsafe_allow_html= */ true);
                }
                let _ctx = col_next;
                {
                    if current_slide < (IntroTour.SLIDES.len() - 1) {
                        if streamlit.button("Next →".to_string(), /* use_container_width= */ true) {
                            streamlit.session_state[IntroTour.SLIDE_KEY] += 1;
                            streamlit.rerun();
                        }
                    } else if streamlit.button("Finish 🎉".to_string(), /* use_container_width= */ true, /* help= */ "Complete the tour and start using ZEN_RAG".to_string()) {
                        IntroTour.mark_intro_completed();
                        streamlit.rerun();
                    }
                }
                let _ctx = col_skip;
                {
                    if streamlit.button("Skip".to_string(), /* use_container_width= */ true) {
                        IntroTour.mark_intro_completed();
                        streamlit.rerun();
                    }
                }
                streamlit.markdown("---".to_string());
            }
        }
    }
}

/// Show intro management controls in sidebar (for testing/reset).
pub fn show_intro_in_sidebar() -> () {
    // Show intro management controls in sidebar (for testing/reset).
    let _ctx = streamlit.sidebar;
    {
        let _ctx = streamlit.expander("ℹ️ Help & Info".to_string(), /* expanded= */ false);
        {
            streamlit.caption("**Intro Tour**".to_string());
            if streamlit.session_state.get(&IntroTour.STATE_KEY).cloned().unwrap_or(false) {
                if streamlit.button("🔄 Show Intro Again".to_string(), /* use_container_width= */ true) {
                    IntroTour.reset_intro();
                    streamlit.rerun();
                }
            } else {
                streamlit.info("Intro tour will display on next visit.".to_string());
            }
            streamlit.markdown("---".to_string());
            streamlit.caption("📚 **Resources**".to_string());
            streamlit.markdown("\n            - [GitHub Repository](https://github.com/GeoHaber/ZEN_RAG)\n            - [Documentation](https://github.com/GeoHaber/ZEN_RAG#readme)\n            - [Issues & Feedback](https://github.com/GeoHaber/ZEN_RAG/issues)\n            ".to_string());
        }
    }
}
